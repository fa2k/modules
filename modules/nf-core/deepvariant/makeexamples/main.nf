process DEEPVARIANT_MAKEEXAMPLES {
    tag "$meta.id"
    label 'process_high'

    //Conda is not supported at the moment
    container "nf-core/deepvariant:1.5.0"

    input:
    tuple val(meta), path(input), path(index), path(intervals)
    tuple val(meta2), path(fasta)
    tuple val(meta3), path(fai)
    tuple val(meta4), path(gzi)

    output:
    // Output a unique ID (hash) for this task, that can be used to join the gvcf and variant calls later.
    tuple val("${task.hash}"), val(meta), val("${prefix}.examples.tfrecord@${task.cpus}.gz"), path("${prefix}.examples.tfrecord-*-of-*.gz"),    emit: examples
    tuple val("${task.hash}"), val(meta), val("${prefix}.gvcf.tfrecord@${task.cpus}.gz"), path("${prefix}.gvcf.tfrecord-*-of-*.gz"),            emit: gvcf
    path "versions.yml",  emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    // Exit if running this module with -profile conda / -profile mamba
    if (workflow.profile.tokenize(',').intersect(['conda', 'mamba']).size() >= 1) {
        error "DEEPVARIANT module does not support Conda. Please use Docker / Singularity / Podman instead."
    }
    def args = task.ext.args ?: ''
    prefix = task.ext.prefix ?: "${meta.id}"
    def regions = intervals ? "--regions=${intervals}" : ""

    """
    seq 0 ${task.cpus - 1} | parallel -q --halt 2 --line-buffer /opt/deepvariant/bin/make_examples \\
        --mode=calling \\
        --ref=${fasta} \\
        --reads=${input} \\
        --examples "./${prefix}.examples.tfrecord@${task.cpus}.gz" \\
        --channels "insert_size" \\
        --gvcf "./${prefix}.gvcf.tfrecord@${task.cpus}.gz" \\
        ${regions} \\
        ${args} \\
        --task {}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        deepvariant_makeexamples: \$(echo \$(/opt/deepvariant/bin/run_deepvariant --version) | sed 's/^.*version //; s/ .*\$//' )
    END_VERSIONS
    """
}
