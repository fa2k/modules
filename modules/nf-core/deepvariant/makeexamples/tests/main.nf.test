nextflow_process {

    name "Test Process DEEPVARIANT_MAKEEXAMPLES"
    script "../main.nf"
    process "DEEPVARIANT_MAKEEXAMPLES"

    tag "deepvariant/makeexamples"
    tag "deepvariant"
    tag "modules"
    tag "modules_nfcore"

    test("homo_sapiens - [bam, bai] - fasta - fai") {
        when {

            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                      file(params.modules_testdata_base_path + '/genomics/homo_sapiens/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true),
                      file(params.modules_testdata_base_path + '/genomics/homo_sapiens/illumina/bam/test.paired_end.sorted.bam.bai', checkIfExists: true),
                      []
                ]
                input[1] = [
                    [ id:'genome'],
                      file(params.modules_testdata_base_path + '/genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                ]
                input[2] = [
                    [ id:'genome'],
                      file(params.modules_testdata_base_path + '/genomics/homo_sapiens/genome/genome.fasta.fai', checkIfExists: true)
                ]
                input[3] = [
                    [],[]
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                // Test string values and versions. The tfrecords are not stable.
                { assert process.out.examples.get(0)[0..1] == [[ id:'test', single_end:false ], "test.examples.tfrecord@2.gz"] },
                { assert process.out.gvcf.get(0)[0..1] == [[ id:'test', single_end:false ], "test.gvcf.tfrecord@2.gz"] },
                { assert snapshot(process.out.versions).match("test1-versions") },
            )
        }
    }

    test("homo_sapiens - [cram, crai, genome_bed] - fasta - fai") {

        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                      file(params.modules_testdata_base_path + '/genomics/homo_sapiens/illumina/cram/test.paired_end.sorted.cram', checkIfExists: true),
                      file(params.modules_testdata_base_path + '/genomics/homo_sapiens/illumina/cram/test.paired_end.sorted.cram.crai', checkIfExists: true),
                      file(params.modules_testdata_base_path + '/genomics/homo_sapiens/genome/genome.bed', checkIfExists: true)
                ]
                input[1] = [
                    [ id:'genome'],
                      file(params.modules_testdata_base_path + '/genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                ]
                input[2] = [
                    [ id:'genome'],
                      file(params.modules_testdata_base_path + '/genomics/homo_sapiens/genome/genome.fasta.fai', checkIfExists: true)
                ]
                input[3] = [
                    [],[]
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert process.out.examples.get(0)[0..1] == [[ id:'test', single_end:false ], "test.examples.tfrecord@2.gz"] },
                { assert process.out.gvcf.get(0)[0..1] == [[ id:'test', single_end:false ], "test.gvcf.tfrecord@2.gz"] },
                { assert snapshot(process.out.versions).match("test2-versions") },
            )
        }
    }

    test("homo_sapiens - [bam, bai] - fasta_gz - fasta_gz_fai") {
        when {

            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                      file(params.modules_testdata_base_path + '/genomics/homo_sapiens/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true),
                      file(params.modules_testdata_base_path + '/genomics/homo_sapiens/illumina/bam/test.paired_end.sorted.bam.bai', checkIfExists: true),
                      []
                ]
                input[1] = [
                    [ id:'genome'],
                      file(params.modules_testdata_base_path + '/genomics/homo_sapiens/genome/genome.fasta.gz', checkIfExists: true)
                ]
                input[2] = [
                    [ id:'genome'],
                      file(params.modules_testdata_base_path + '/genomics/homo_sapiens/genome/genome.fasta.gz.fai', checkIfExists: true)
                ]
                input[3] = [
                    [ id:'genome'],
                      file(params.modules_testdata_base_path + '/genomics/homo_sapiens/genome/genome.fasta.gz.gzi', checkIfExists: true)
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert process.out.examples.get(0)[0..1] == [[ id:'test', single_end:false ], "test.examples.tfrecord@2.gz"] },
                { assert process.out.gvcf.get(0)[0..1] == [[ id:'test', single_end:false ], "test.gvcf.tfrecord@2.gz"] },
                { assert snapshot(process.out.versions).match("test3-versions") },
            )
        }
    }

}
