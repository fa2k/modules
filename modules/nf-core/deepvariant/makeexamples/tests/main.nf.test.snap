{
    "test3-versions": {
        "content": [
            [
                "versions.yml:md5,842dca9323f25aa3cfd67789d18e7e33"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-08-09T11:38:03.846393488"
    },
    "test2-versions": {
        "content": [
            [
                "versions.yml:md5,842dca9323f25aa3cfd67789d18e7e33"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-08-09T11:37:54.085974225"
    },
    "test1-versions": {
        "content": [
            [
                "versions.yml:md5,842dca9323f25aa3cfd67789d18e7e33"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-08-09T11:37:44.424124268"
    }
}